File: code/captsone_project/frontend/travel/src/app/core/services/booking.service.ts
import { Injectable } from '@angular/core';
import { Observable, timer } from 'rxjs';
import { switchMap, takeWhile } from 'rxjs/operators';
import { ApiService } from './api.service';
import { environment } from '../../../environments/environment';
import { Booking } from '../models/booking.model';
import { Passenger } from '../models/passenger.model';

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  constructor(private api: ApiService) {}

  createBooking(flightId: number, passengers: Passenger): Observable<Booking> {
    return this.api.post<Booking>(environment.endpoints.createBooking, passengers, { flightId });
  }

  getBookingById(id: number): Observable<Booking> {
    return this.api.get<Booking>(environment.endpoints.bookingById(id));
  }

  getStatus(id: number): Observable<{ status: string }> {
    return this.api.get<{ status: string }>(environment.endpoints.bookingStatus(id));
  }

  pollStatus(id: number, intervalMs = 1000, timeoutMs = 30000): Observable<{ status: string }> {
    const start = Date.now();
    return timer(0, intervalMs).pipe(
      switchMap(() => this.getStatus(id)),
      takeWhile(() => Date.now() - start < timeoutMs, true)
    );
  }
}

File: code/captsone_project/backend/booking-ms/src/main/java/com/wipro/bookingms/service/BookingService.java
package com.wipro.bookingms.service;

import com.wipro.bookingms.model.Booking;
import com.wipro.bookingms.model.Passenger;
import java.util.List;

import java.util.Map;

public interface BookingService {
    Booking createBooking(Long flightId, List<Passenger> passengers);

    void initiatePayment(Long bookingId, Map<String, Object> paymentDetails);

    Booking getBookingById(Long id);

    String getBookingStatus(Long id);
}